### PR TITLE
Fix root path being truncated when formatting errors

### LINKED ISSUE #14725: C# compiler is not working properly when called from root folder
**Version Used**: Microsoft (R) Visual C# Compiler version 2.0.0.60925

**Steps to Reproduce**:
1. Go to root folder (e.g. c:\ or /)
2. Compile any C# code with syntax error using full path to the file

C:>csc c:\temp\p.cs
Microsoft (R) Visual C# Compiler version 2.0.0.60925
Copyright (C) Microsoft Corporation. All rights reserved.

emp\p.cs(6,24): error CS1003: Syntax error, ',' expected

**Observe missing c:\t**

### PR DESCRIPTION
**Pull request by OmarTawfik**
Fixes #14725 
@dotnet/roslyn-compiler

### CODE REVIEW

File: src/Compilers/CSharp/Portable/CommandLine/CommandLineDiagnosticFormatter.cs [hunk 1]
```diff
@@ -66,7 +66,9 @@ internal string RelativizeNormalizedPath(string normalizedPath)
             var directory = PathUtilities.GetDirectoryName(normalizedPath);
             if (string.Compare(directory, 0, normalizedBaseDirectory, 0, normalizedBaseDirectory.Length, StringComparison.OrdinalIgnoreCase) == 0)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/CommandLine/CommandLineDiagnosticFormatter.cs]: It looks like this check does not handle `baseDirectory: "C:\AB"` and `normalizedPath: "C:\ABCD\file.cs"`. Should it?

File: src/Compilers/CSharp/Portable/CommandLine/CommandLineDiagnosticFormatter.cs [hunk 2]
```diff
@@ -66,7 +66,9 @@ internal string RelativizeNormalizedPath(string normalizedPath)
...
             {
-                return normalizedPath.Substring(normalizedBaseDirectory.Length + 1);
+                return PathUtilities.IsDirectorySeparator(normalizedBaseDirectory.Last())
+                    ? normalizedPath.Substring(normalizedBaseDirectory.Length)
+                    : normalizedPath.Substring(normalizedBaseDirectory.Length + 1);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/CommandLine/CommandLineDiagnosticFormatter.cs]: Consider moving the `IsDirectorySeparator` call inside `Substring`:
```
return normalizedPath.SubString(IsDirectorySeparator(...) ? Length : Length + 1);
```

File: src/Compilers/CSharp/Portable/CommandLine/CommandLineDiagnosticFormatter.cs (hunk 1 above)
Comment (OmarTawfik) [inline src/Compilers/CSharp/Portable/CommandLine/CommandLineDiagnosticFormatter.cs]: Great catch!
Fixed it and added a test as well.

**Comment by AlekseyTs** [comment]
Does VB have similar issue.

File: src/Compilers/CSharp/Test/CommandLine/CommandLineDiagnosticFormatterTests.cs [hunk 3]
```diff
@@ -28,5 +28,39 @@ public void GetPathNameRelativeToBaseDirectory()
...
+        [ConditionalFact(typeof(WindowsOnly))]
+        public void RelativizeNormalizedPathShouldHandleDirectoriesWithSamePrefix_2()
+        {
+            var formatter = new CommandLineDiagnosticFormatter(
+                baseDirectory: @"C:\ABCD",
+                displayFullPaths: false,
+                displayEndLocations: false);
+
+            Assert.Equal(@"c:\AB\file.cs", formatter.RelativizeNormalizedPath(@"c:\AB\file.cs"));
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineDiagnosticFormatterTests.cs]: Please also test sub-directories between `baseDirectory` and `normalizedPath`. e.g.: `baseDirectory: "C:\A\B"`, `normalizedPath: "C:\A\B\C\D\file.cs"`

File: src/Compilers/Core/Portable/FileSystem/PathUtilities.cs [hunk 4]
```diff
@@ -107,6 +107,25 @@ internal static string GetDirectoryName(string path, bool isUnixLike)
...
+        {
+            parent = RemoveTrailingDirectorySeparator(parent);
+
+            while (child != null)
+            {
+                child = RemoveTrailingDirectorySeparator(child);
+
+                if (child.Equals(parent, StringComparison.OrdinalIgnoreCase))
+                {
+                    return true;
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/FileSystem/PathUtilities.cs]: >return true; [](start = 20, length = 12)

It looks like we will get here if the function is called with equal arguments. Is this what we want?

**Comment by OmarTawfik** [comment]
@AlekseyTs VB always print the full paths.

File: src/Compilers/Core/Portable/FileSystem/PathUtilities.cs (hunk 4 above)
Comment (OmarTawfik) [inline src/Compilers/Core/Portable/FileSystem/PathUtilities.cs]: Yes. I'll rename to remove ambiguity.

**Comment by OmarTawfik** [comment]
@AlekseyTs @cston do I've sign offs?
/cc @dotnet/roslyn-compiler

**Comment by cston** [comment]
LGTM

File: src/Compilers/Core/Portable/FileSystem/PathUtilities.cs [hunk 5]
```diff
@@ -107,6 +107,25 @@ internal static string GetDirectoryName(string path, bool isUnixLike)
             return null;
         }
 
+        internal static bool IsDirectorySameOrChildOf(string child, string parent)
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/FileSystem/PathUtilities.cs]: >IsDirectorySameOrChildOf [](start = 29, length = 24)

IsSameDirectoryOrChildOf ?

**Comment by AlekseyTs** [comment]
LGTM

File: src/Compilers/CSharp/Test/CommandLine/CommandLineDiagnosticFormatterTests.cs (hunk 3 above)
Comment (marek-safar) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineDiagnosticFormatterTests.cs]: You should also write some non-windows specific tests as the issue was present on non-windows platform as well

**Pull request merged**